Add developing with hydra guide (#433)
input-output-hk/hydra: docs/docs/getting-started/developing-on-hydra.md

# Hydra development

This guide is meant to be a tutorial on how to develop an application on Hydra. It will assume the reader is already familiar with developing a DApp on Cardano and will focus on the differences between the two.

## On-chain code

> On-chain code will be exactly the same between Cardano and a hydra head. This is one of the main selling points of the hydra design (heads are also called isomorphic state channels) the only caveat being that (at the time of writing 07-22) hydra heads do not support validity ranges for transactions, the reason for this is there is currently no notion of time inside of a hydra head.

Reviewer: ```suggestion
On-chain code will be exactly the same between Cardano and a hydra Head. This is one of the main selling points of the hydra design (heads are also called isomorphic state channels) the only caveat being that (at the time of writing 07/22) Hydra Heads do not support validity ranges for transactions, the reason for this is there is currently no notion of time inside of a Hydra Head. See also this [feature on the roadmap](https://github.com/input-output-hk/hydra-poc/issues/196)
```

## Off-chain code

Here is where the differences will begin. Usually, DApp developers will write some off-chain code to interact with their validators and minting policies.
This off-chain code will usually be exposed through an HTTP interface that frontends can access to drive interactions with the application forward.

> Off-chain code is usually written in the Contract monad, which is used to build transactions and interact with the Cardano blockchain in a controlled way.

Reviewer: I don't the that the PAB is THE standard way of writing DApps. It might be that readers are familiar with it if they have taken part in the plutus pioneers program.

Can you make this a bit less opinionated? i.e. state that it's not like in the PPP because we there is also no PAB support (yet, there is also this ticket: https://github.com/input-output-hk/hydra-poc/issues/214 and has not been tackled because nobody shouted as us yet + the PAB is kind of getting redesigned).

When it comes to hydra, the developer will not have access to similar tools. The interface to speak to a hydra node is a websocket. Developers must build applications that connect to these websockets and can send and receive events through them.

The Hydra node will emit several commands through this websocket: the API is defined [here](https://hydra.family/head-protocol/api-reference).
The API reference is divided into two sections: Pub and Sub.

> The first one describes the commands that a hydra node will accept, while the second one describes the events that will be emitted through the websocket.

Reviewer: Before diving into the API endpoints (below), is there something to consider on initializing & opening the head?

> An important event developers will be interested in is `SnaphotConfirmed`. This event is emitted once all the hydra nodes that are partaking in the head, have seen and signed a transaction submitted by one of them.

Reviewer: ```suggestion
An important event developers will be interested in is `SnaphotConfirmed`. This event is emitted once all the participants of the head, have seen and signed a transaction submitted by one of them.
```

> Once a transaction lands inside a snapshot, it is confirmed and there are no chances for it to be rolled back. This is an important distinction from L1: in that case, when a node sees a transaction there is still a chance that it might be rolled back (where the probability of this goes to 0 as more blocks are added), in the hydra world, the only way to make a snapshot is to have every single node in the head approve of it.

Reviewer: About rollbacks: this is only true after the head has been open long enough and the opening tx has become final. Might want to link this feature for more details: https://github.com/input-output-hk/hydra-poc/issues/185

This is a synchronous process and requires all the head participants to be online all the time, the tradeoff is that there is no uncertainty or possibility of rollbacks, the moment the snapshot is confirmed there is no going back.

> For the Pub side of the API, an important endpoint will be `NewTx`. This is used to submit a transaction to the node. There are a couple of ways the transaction can be submitted, but the simplest method we found is to simply serialise the transaction built with `Cardano.Api` to CBOR and submit that through a websocket.

Reviewer: ```suggestion
For the Pub side of the API, an important endpoint will be `NewTx`. This is used to submit a transaction to the node. There are a couple of ways the transaction can be submitted, but simple and familiar way should be to build the transaction with `Cardano.Api`, sign & serialize it to CBOR and submit that through the websocket.
```

## Testing

> Testing inside the hydra head can be quite different than for normal DApps, the reason is you will not have access to EmulatorTrace or an easy way to spin up a dev environment to test.

Reviewer: ```suggestion
Testing inside the hydra head can be quite different than for workign with the PAB, the reason is you will not have access to EmulatorTrace or a pre-deployed playground environment to test.
```

> However it is possible to spin up a cardano-node, and a cluster of hydra-nodes programmatically and run tests directly on that.

Reviewer: ```suggestion
Instead, it is possible to spin up a local cardano devnet, as well as a cluster of hydra-nodes programmatically and run tests directly on that.
```

> The module hydra-cluster provides functions to spin up the set of required nodes and to run assertions on their output.

Reviewer: I have not considered that `hydra-cluster` would be used downstream, but that is actually an interesting view on it :D

```suggestion
The hydra-cluster package is used within Hydra to do end-to-end tests on the protocol and hence is a nice blueprint for this setup. It's library component provides functions to spin up the required nodes, send API inputs and run assertions on the API output. Should you be missing something or need more configurability, feel free to open a feature request and we are happy to support you.
```

A couple of steps are required to set up the initial environment:

> - We need to spin up at least one Cardano BFT node. This can be done with `withBFTNode` which takes the cardano node configuration and a callback that has access to an instance of the running node. The node parameters can be generated by a helper also exported by hydra-cluster: `newNodeConfig`.

Reviewer: Quite some things will / have been renamed in #430. I'll suggest the new names:

```suggestion
- We want to spin up a local Cardano devnet comprised by a single block producer. This can be done with `withCardanoDevnet` which takes the a working directory where blocks and logs will be stored, as well as a callback that gets access to the `RunningNode`. Note that the devnet will use credentials and configuration from the `hydra-cluster/config/`.
```

- In the callback, we can then spin up a cluster of hydra-nodes using `withHydraCluster`. This will require the cardano-node socket, which can be obtained from the running node instance, and a set of cardano and hydra keys which can also be easily generated through helpers exposed by `hydra-cluster`.

> Let's walk through a slightly simplified version of one of our tests, with comments explaining the setup we just described.

Reviewer: ```suggestion
Let's walk through a simple example, with comments explaining the setup we just described.
```

This test will spin up a single cardano node, and a pair of hydra nodes (Alice and Bob).

```haskell
-- Tracer is an instance of https://hackage.haskell.org/package/contra-tracer and is used by several
-- hydra components to deal with logs in a structured way

> it "spins up two hydra nodes" $ \tracer ->

Reviewer: This would not be in scope, maybe we should suggest to use a `let tracer = stdoutTracer` here.

  -- Set a timeout for the whole test to fail
  failAfter 60 $
    -- We create a temporary directory to contain all the files required to spin up a node

>     withTempDir "cardano-node-tmp-dir" $ \cTmpDir -> do
>       config <- newNodeConfig cTmpDir
>       -- This starts the BFTNode as described above
>       withBFTNode (contramap FromCardanoNode tracer) config $ \(RunningNode _ nodeSocket) -> do

Reviewer: ```suggestion
    withTempDir "hydra-cluster-example-tmp-dir" $ \tmpDir -> do
      -- This starts the cardano devnet as described above
      withCardanoDevnet (contramap FromCardanoNode tracer) tmpDir $ \RunningNode{nodeSocket} -> do
```

        -- We create a temporary directory to contain all the files required to spin up the hydra nodes

>         withTempDir "hydra-nodes-tmp-dir" $ \hTmpDir -> do

Reviewer: I would recommend to re-use the temp dir to make this simpler. No point in separating, both will now (soon #430) write to `logs/cardano-node.log` and `logs/hydra-node-x.log`.

          -- We generate pairs of cardano keys for alice and bob, this pair is a tuple of verification and signing key
          aliceKeys <- generate genKeyPair
          bobKeys <- generate genKeyPair

          -- Here we generate signing keys for hydra, note these are different from the cardano skeys generated above
          let aliceSk = generateSigningKey "alice"
              bobSk = generateSigningKey "bob"

          let cardanoKeys = [aliceKeys, bobKeys]
              hydraKeys = [aliceSk, bobSk]

          -- Hydra nodes take an id as one of their arguments, it is used to identify them with the head protocol
          let firstNodeId = 0

          -- This will spin up `n` hydra nodes, where `n` is the length of the cardanoKeys list
          -- (which needs to match the length of hydraKeys)
          withHydraCluster tracer hTmpDir nodeSocket firstNodeId cardanoKeys hydraKeys $ \nodes -> do

>             ...

Reviewer: If you do a `pure ()` here, we could even think of making this whole thing doctestable / compilable?

I think this is missing distributing funds with the `seedFromFaucet` stuff and maybe we also want to include some actual `send "Init" ["contestationPeriod" .= 10]` etc.?

Your text below could become comments in between.

```

At this point, we are ready to start sending commands to the hydra-nodes programmatically.
Hydra-cluster also exposes a `send` function, which takes an instance of the hydra-node and uses its websocket connection to send commands to the process.
Remember that we must handle the whole head initialisation inside the test as well, so the first step will always be for one of the two nodes to `send` an `Init` command
to start the opening of the head.

> Finally, hydra-node exposes several useful functions to wait for output from the hydra nodes themselves, both these functions have a timeout so they will function as assertions from a testing point of view; that is, if the node does not output what we expect within a certain timeframe, an error will be thrown and the whole test will fail.

Reviewer: There should not be anything to import from `hydra-node`, what were you using from `hydra-node`?

Author: That was a typo :( meant `hydra-cluster`

Remember the nodes will always produce output in JSON format, so all the assertions will expect some form of Aeson.Value to check against.

`waitFor` will take a list of hydra-nodes and an Aeson Value, and check if they all produce the expected output within the allowed timeframe.

`waitMatch` instead can be used to parse the output and possibly extract some information from that to be used in subsequent steps of the test.

> For example, after we submit a transaction to the node, we can use `waitMatch` to parse the `SnaphotConfirmed` event and extract the transactions present in that snapshot to check if it contains the transaction we just sent.

Reviewer: Instead of documenting the functions here, probably better to just link the haddocks (https://hydra.family/head-protocol/haddock/hydra-cluster/HydraNode.html) and enumerate them as handy functions. 

Should there be no haddocks -> move these docs there :)
